codes/python/chapter_array_and_linkedlist/my_list.py

```diff
@@ -4,75 +4,82 @@
 Author: Krahets (krahets@163.com)
 """
 
-import sys, os.path as osp
-sys.path.append(osp.dirname(osp.dirname(osp.abspath(__file__))))
-from include import *
+from typing import List
 
 """ 列表类简易实现 """
 class MyList:
-    """ 构造方法 """
     def __init__(self):
-        self.__capacity = 10                 # 列表容量
-        self.__nums = [0] * self.__capacity  # 数组（存储列表元素）
-        self.__size = 0                      # 列表长度（即当前元素数量）
-        self.__extend_ratio = 2              # 每次列表扩容的倍数
+        """ 构造方法 """
+        self.__capacity = 10  # 列表容量
+        self.__nums: List[int] = [0] * self.__capacity  # 数组（存储列表元素）
+        self.__size = 0  # 列表长度（即当前元素数量）
+        self.__extend_ratio = 2  # 每次列表扩容的倍数
 
-    """ 获取列表长度（即当前元素数量） """
-    def size(self):
+    def __check_index(self, index: int):
+        """检查索引是否合法"""
+        assert 0 <= index < self.__size, "索引越界: size-> {}, index->{}".format(self.__size, index)
+
+    def __check_size_or_extend_capacity(self):
+        if self.__size == self.__capacity:
+            self.__extend_capacity()
+
+    def size(self) -> int:
+        """ 获取列表长度（即当前元素数量） """
         return self.__size
-    
-    """ 获取列表容量 """
-    def capacity(self):
+
+    def capacity(self) -> int:
+        """ 获取列表容量 """
         return self.__capacity
-    
-    """ 访问元素 """
-    def get(self, index):
+
+    def get(self, index: int) -> int:
+        """ 访问元素 """
         # 索引如果越界则抛出异常，下同
-        assert index >= 0 and index < self.__size, "索引越界"
+        self.__check_index(index)
         return self.__nums[index]
 
-    """ 更新元素 """
-    def set(self, num, index):
-        assert index >= 0 and index < self.__size, "索引越界"
-        self.__nums[index] = num
-
-    """ 中间插入（尾部添加）元素 """
-    def add(self, num, index=-1):
-        assert index >= 0 and index < self.__size, "索引越界"
-        # 若不指定索引 index ，则向数组尾部添加元素
-        if index == -1:
-            index = self.__size
-        # 元素数量超出容量时，触发扩容机制
-        if self.__size == self.capacity():
-            self.extend_capacity()
-        # 索引 i 以及之后的元素都向后移动一位
+    def set(self, value: int, index: int) -> None:
+        """ 更新元素 """
+        self.__check_index(index)
+        self.__nums[index] = value
+
+    def add(self, value: int) -> None:
+        """尾部插入元素"""
+        self.__check_size_or_extend_capacity()
```

Comment:
Please do not add a new function because all the codes are synchronized among 12 languages. Change other language all if you want to change the Python code.